storybookjs/storybook
# Vite: Replace vite-plugin-externals with custom plugin

Issue: #20643



## What I did

* When used in dev mode, `vite-plugin-externals` breaks HMR for vite-based projects.
* When disabled in dev mode, it completely prevents Storybook from working on vite + pnpm (and probably Yarn PnP) projects.
* I replaced that plugin with a local one, which is straightforward and quite naive

## What I'm not sure I checked correctly

* Production build works fine
* That the #20514 issue about code duplication is still fixed (I've searched for 

## What I'm sure I didn't checked
* Yarn PnP

## How to test

- You can create any Vite sandbox, but it will still require you to change the package manager to pnpm so that dependencies do not get hoisted.
- Make sure you get the following error: `[vite] Internal server error: Failed to resolve import "@storybook/preview-api" from "../../../../../../virtual:/@storybook/builder-vite/vite-app.js". `
- Build the `@storybook/builder-vite` package from the source branch of this PR
- Copy the generated `dist` output of `@storybook/builder-vite` to replace the one in your pnpm workspace 
- Start storybook
- Be happy because the error is gone!

## Checklist



      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

      make sure to add the `ci:merged` or `ci:daily` GH label to it.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`
